src/test/playwright/service-worker.spec.ts

```diff
@@ -0,0 +1,286 @@
+          `Found ${versionMatches.length}.`
+      );
+    }
+    const hex = versionMatches[0].groups?.hex;
+    if (!hex) {
+      throw new Error('Could not find "hex" capture group in regexp match.');
+    }
+    originalSwVersion = hex;
+
+    const plugin: Plugin = {
+      name: 'test-plugin',
+
+      serve(ctx) {
+        if (ctx.path === '/') {
+          return indexHtml;
+        }
+        return undefined;
+      },
+
+      transform(ctx) {
+        let body = ctx.body as string;
+        if (typeof body !== 'string') {
+          throw new Error('Expected string body');
+        }
+        if (ctx.path.endsWith('playground-service-worker.js')) {
+          body = swCode
+            .replace(originalSwVersion, swVersion)
+            // Extreme hack! We need some way to check which service worker
+            // served a response. Prepend the version into all responses.
+            .replace(/new Response\(/g, `new Response("${swVersion} "+`);
+          serviceWorkerServedDeferred.resolve();
+          serviceWorkerServedDeferred = new Deferred();
+        } else {
+          body = body.replace(originalSwVersion, expectedVersion);
+        }
+        return {body};
+      },
+
+      transformCacheKey() {
+        // By default @web/dev-server will cache transformed responses in memory
+        // based purely on the request path. We need to override this because
+        // [1] we serve varying responses for the same path, and [2] we want to
+        // observe every response so we can resolve a promise.
+        return String(Math.random());
+      },
+    };
+
+    server = await startDevServer({
+      config: {
+        rootDir: '.',
+        nodeResolve: true,
+        plugins: [plugin],
+      },
+    });
+    indexUrl = `http://localhost:${server.config.port}/`;
+  });
+
+  test.beforeEach(() => {
+    setServiceWorkerCode(originalSwCode);
+    setServiceWorkerAndExpectedVersion(originalSwVersion);
+    serviceWorkerServedDeferred = new Deferred();
+  });
+
+  test.afterAll(async () => {
+    await server.stop();
+  });
+
+  const getIframeBodyText = async (page: Page): Promise<string> => {
+    const iframe = await page.waitForSelector('playground-preview iframe');
+    const contentFrame = await iframe.contentFrame();
+    if (contentFrame === null) {
+      throw new Error('iframe had null contentFrame');
+    }
+    const bodyText = await contentFrame.textContent('body');
+    if (bodyText === null) {
+      throw new Error('contentFrame had null body text');
+    }
+    return bodyText;
+  };
+
+  const expectIframeServesVersion = async (
+    page: Page,
+    version: string
+  ): Promise<void> => {
+    const text = await getIframeBodyText(page);
+    expect(text).toEqual(`${version} index.html`);
+  };
+
+  const clickReloadButton = async (page: Page): Promise<void> => {
+    await page.click('#reload-button');
+  };
+
+  const expectConsoleLog = (page: Page, expect: string): Promise<void> =>
```

Comment:
Does this test have a fail state? I.e. if there is a console message that doesn't match -> reject.

I'm assuming the fail state results in a timeout?

Reply:
Yeah it just relies on the timeout to fail. There are various other things in the console such as the "lit is in dev mode!" message, so we can't just wait for the next message that appears (plus even if we did, we'd still have to rely on a timeout to catch the case where it never logs anything).